add new guide (#1120)
mattermost/mattermost-developer-documentation: site/content/contribute/why-contribute/_index.md

---
title: "Section 1 — Why you might contribute"
heading: "Section 1 — Why you might contribute"

> description: ""

Reviewer: Thanks for removing this, @neflyte!

date: "2022-10-01T12:00:00-00:00"
weight: 1
---

## Subsection 1 — You’ve found a bug

1. If the bug fix that you’re proposing would be larger than 20 lines of code, [create a GitHub issue](https://github.com/mattermost/mattermost-server/issues).
    1. You can speed up the process by asking about the issue in the [Contributors](https://community.mattermost.com/core/channels/tickets) or [Developers](https://community.mattermost.com/core/channels/developers) channels.
    2. [Here’s a good example of a ticket](https://github.com/mattermost/mattermost-mobile/pull/364/commits/7a97451b62fee4022edac4c0395ad0a5cbf1bb66) that is small enough to not need a ticket while still being incredibly helpful.
2. If you’ve volunteered to take the ticket once its active, or if your fix is too small to warrant the ticket, fork the applicable repo and start making your changes.

> 3. Create a PR on the `master` branch of the applicable repo. If there is an associated [Jira](https://mattermost.atlassian.net/issues/?jql=) or GitHub issue, your PR should begin with the issue ID (e.g. `[MM-394]` or `[GH-394]`). Our GitHub PR template will walk you through the creation of your PR.

Reviewer 2: Link to GitHub PR template from last line:
"Our GitHub PR template" should ideally point to somewhere.
I know Ben worked on a template, but not sure if his templates are already available - https://docs.google.com/document/d/11uubL-_RBEML6r3yOmDRCcTyT8CAJljiKjNdr56kpDg/edit

Reviewer: @azigler - Do you know if there's Ben's GitHub templates are formalized such that we could point readers somewhere specific as per Jordana's feedback?

Reviewer 3: I'm pretty sure those templates are part of what he wants to give developers when they spin off an open source library. I wasn't involved in making them so I can't weigh in on how "formalized" they are, unfortunately.

Reviewer: @jordanafung - I've noted this as a follow-up task we can take action on in a future iteration.

        1. Generally avoid doing QA testing yourself, unless it’s obvious that it’s not necessary. In any case, it likely won’t be obvious to onlookers, so clearly explain the rationale in the comments on the PR.
    6. Once all outstanding requests are completed and all involved have approved of the PR, add the `4: Reviews Complete` if it hasn’t been already, merge the PR into `master`, and delete the old branch. If your pull request depends on other pull requests, consider assigning the `Do Not Merge/Awaiting PR` label to avoid merging prematurely.
    7. If this is a cherry-picked PR, the automation should handle everything from here on out. If the automated cherry-pick fails, you will need to cherry-pick the PR manually. Cherry-pick the master commit back to the appropriate releases. If the release branches have not been cut yet, leave the labels as-is and cherry-pick once the branch has been cut. The release manager will remind you to finish your cherry-pick. Set the `CherryPick/Done` label when you’re done with this process.

>     8. Update the appropriate Jira ticket so everybody knows where the project stands. Resolve the ticket for QA from “Ready for QA” button with QA test steps, or with “No Testing Required” if no QA testing is needed.

Reviewer 4: Would a GitHub ticket also be updated in this case?

Reviewer: @neflyte - This content applies to Mattermost staff rather than community contributors. Let's revisit this section together once outstanding technical and editorial feedback and reviews are incorporated but before we merge this PR.

    9. If you need to test your changes on a test server, you can add the appropriate label to the PR and wait 3-5 minutes for the server to be created (see [the labels page]({{< ref "/contribute/more-info/getting-started/labels.md)" >}}). A bot will send you your link and credentials in the form of a comment on the PR. Remove the label when you’re done to destroy the test server.
6. Once you address suggestions a reviewer has made, re-request a review from them. Their initial review was technically completed, so it’s no longer in their queue until you re-request.

>     1. Give reviewers time — they have long queues and different schedules. If you’ve been assigned a reviewer but haven’t heard from them in 5 business days, politely mention them in a PR comment.

Reviewer 4: ```suggestion
    1. Give reviewers time — they may have long queues and different schedules. If you’ve been assigned a reviewer but haven’t heard from them in 5 business days, politely mention them in a PR comment.
```

        1. When contributions are inactive but for good reason (for example, when the team is actively discussing a major design decision but they haven’t decided on anything yet), `lifecycle/frozen` would be a better label.
        2. Inactive contributions are eligible to be picked up by another community member.
    3. After 30 days of inactivity, a contribution becomes orphaned, and they’ll add the `lifecycle/3:orphaned` label to the now-closed PR. The associated `Help Wanted` ticket is given back its `Up For Grabs` status so others can pick up on the issue.

## Subsection 2 — You want to help with content

1. Good content is just as important as good code! If you notice and fix a content error in the documentation or in another open source article describing Mattermost, we consider you to be a valued member of our contributor community just like if you had added a feature to the core code.

> 2. If you see a problem with Mattermost [developer](https://developers.mattermost.com/) or [user](https://docs.mattermost.com/) documentation, you have a few options:

Reviewer 4: ```suggestion
2. If you see a problem with Mattermost [developer](https://developers.mattermost.com/) or [product](https://docs.mattermost.com/) documentation, you have a few options:
```

    1. If you have time to fix the mistake and it only affects a single page, navigate to the applicable page and click “Edit in GitHub” at the top right.  It’ll walk you through the process of creating a fork so that you can then follow the steps under section 1, subsection 1 of this guide, “You’ve found a bug”.
    2. If you don’t have time to fix the mistake, copy the file path you’re on and create a GitHub issue about the problem you found. Make sure to include the file path and fill out the issue template completely to maximize clarity.

>     3. If you’re not up for creating a GitHub issue right now, that’s alright too! In the bottom right corner of every docs page is the question “Did you find what you were looking for?” Use this to quickly provide direct feedback about any page you’re viewing.

Reviewer 4: This only applies to the product docs at the moment. There is an existing request to add this functionality to the developer docs site. FYI.

>     4. If you want to fix a larger problem that affects multiple pages or the structure of the docs, you should first report it as an issue on the appropriate GitHub repo, and follow the steps under section 1, subsection 1 of this guide, “You’ve found a bug”. The developer and user docs repositories contain instructions on how to build and modify the sites locally so you can test larger changes more efficiently.

Reviewer 4: ```suggestion
    4. If you want to fix a larger problem that affects multiple pages or the structure of the docs, you should first report it as an issue on the appropriate GitHub repo, and follow the steps under section 1, subsection 1 of this guide, “You’ve found a bug”. The developer and product docs repositories contain instructions on how to build and modify the sites locally so you can test larger changes more efficiently.
```

>     5. The best place to discuss problems with us is in the [Documentation Working Group channel](https://community.mattermost.com/core/channels/dwg-documentation-working-group) where you can ping our technical writers with the hashtag #docsteam.

Reviewer 4: Is it `#docsteam` or `@docsteam`? I'm 3/5 on `@docsteam`

Reviewer: The Technical Writing team has a user group called ``@docsteam``. @neflyte - Are you open to updating this PR with this correction?

3. If you’d like to contribute to our blog, website, or social media, you also have a few options:
    1. You can get paid to write technical content for developers through the Mattermost [community writing program](https://mattermost.com/blog/blog-announcing-community-writing-program/).

>     2. If you see a problem with any webpages, blog posts, or other content on [Mattermost.com](http://Mattermost.com), you can notify us via [the Content channel](https://community.mattermost.com/core/channels/mattermost-blog).

Reviewer 4: ```suggestion
    2. If you see a problem with any webpages, blog posts, or other content on [Mattermost.com](https://mattermost.com), you can notify us via [the Content channel](https://community.mattermost.com/core/channels/mattermost-blog).
```

    3. Share your contributor or user experience! Mention us when you promote work you do in our community and we’ll amplify the message through Mattermost social media profiles.
    4. Want to lead a social community? We can provide advice and resources to help you in the [Community channel](https://community.mattermost.com/core/channels/community-team).

## Subsection 3 — You’d like to make something more inclusive or accessible

1. Accessibility is one of the most overlooked yet most important features of modern software development, and we’re eager to improve on the accessibility of the features in our open source project and its documentation.
2. Problems with the accessibility or the inclusivity of both features in the codebase and pieces of content describing Mattermost are bugs, so we treat them as such in our development process. See subsections 1 and 2 for the description on how you can submit your contribution.

> 3. When you contribute a change that incorporates an adjustment based on the principles of accessibility and inclusivity, please use this guide to back you up in the PR, ticket, or post.

Reviewer 4: Why would someone need to back up their PR if it's addressing a shortfall in accessibility and/or inclusivity?

Reviewer: @neflyte - Given the volume of feedback outstanding, let's revisit this content together prior to merging this PR.

    1. Take a look at our [product roadmap](https://mattermost.com/roadmap/). There’s a chance we might already be building the thing you want.
    2. [Provide input on features](https://portal.productboard.com/mattermost/33-what-matters-to-you/tabs/117-potential-future-offerings) we’re considering to let us know what matters the most to you.
    3. [Participate in a survey](https://portal.productboard.com/mattermost/33-what-matters-to-you/tabs/115-help-us-learn-more-through-surveys) to help us better understand how to meet the needs of our users.
    4. Discuss your idea with our community in the [Feature Proposals channel](https://community.mattermost.com/core/channels/feature-ideas).
    5. Build an app! Mattermost has a rich framework full of tools to help you add the features you want that don’t quite work as core additions to Mattermost.
        1. [Webhooks]({{< ref "/integrate/webhooks/" >}}) are the easiest type of app to create and enable you to integrate with the vast majority of modern web tools.
        2. [Slash commands]({{< ref "/integrate/slash-commands/" >}}) are apps that respond to triggers sent as messages inside Mattermost.

>         3. More complicated apps can be built as standard HTTP services in any programming language using a robust API. [Get started here]({{< ref "/integrate/apps/" >}}).

Reviewer 4: ```suggestion
        3. More complicated Apps can be built as standard HTTP services in any programming language using a robust API. [Get started here]({{< ref "/integrate/apps/" >}}).
```
We capitalize the `A` in `Apps` when referring to the Apps framework vs. other uses of the word `app`.

        4. If you’re willing to deal with more complex development overhead, [plugins]({{< ref "/integrate/plugins/" >}}) enable you to alter every component of the Mattermost experience.

## Subsection 5 — You’re looking to practice your skills or give back to the community

> 1. We love developers who are passionate about open-source. If you’re looking to tackle an interesting problem, we’ve got you covered! Feel free to check out [the help wanted tickets on GitHub](https://github.com/search?utf8=%E2%9C%93&q=is%3Aopen+archived%3Afalse+org%3Amattermost+label%3A%22Help+Wanted%22++label%3A%22Up+For+Grabs%22). To take one on, just comment on the issue and follow the process outlined in article 1.1 of this guide.

Reviewer 4: ```suggestion
1. We love developers who are passionate about open-source. If you’re looking to tackle an interesting problem, we’ve got you covered! Feel free to check out [the help wanted tickets on GitHub](https://github.com/search?utf8=%E2%9C%93&q=is%3Aopen+archived%3Afalse+org%3Amattermost+label%3A%22Help+Wanted%22++label%3A%22Up+For+Grabs%22). To take one on, just comment on the issue and follow the process outlined in section 1, sub-section 1 of this guide.
```

Reviewer: @neflyte - Can you please replace the section/subsection reference with a direct link to the section where readers can  see more details?
